### ISSUE TITLE
UnboundLocalError: local variable 'feature_db' referenced before assignment

### ISSUE THREAD

**Issue by Kohei-Fujiwara89** [agshumate/Liftoff#87]
I also have the same error as the issue #75, #86.

After I checked these posts, I confirmed 
	there's no _db file before running,
	the annotation file is .gff3,
	the permission of the annotation file is -rwxrwxrwx,
	and I did umask 0000 as default permission setting.

However, there comes still the similar error like,

Traceback (most recent call last):
  File "/home/usr/anaconda3/envs/liftoff/bin/liftoff", line 10, in <module>
    sys.exit(main())
  File "/home/usr/anaconda3/envs/liftoff/lib/python3.6/site-packages/liftoff/run_liftoff.py", line 10, in main
    run_all_liftoff_steps(args)
  File "/home/usr/anaconda3/envs/liftoff/lib/python3.6/site-packages/liftoff/run_liftoff.py", line 25, in run_all_liftoff_steps
    parent_features_to_lift)
  File "/home/usr/anaconda3/envs/liftoff/lib/python3.6/site-packages/liftoff/liftover_types.py", line 13, in lift_original_annotation
    parents_to_lift, args)
  File "/home/usr/anaconda3/envs/liftoff/lib/python3.6/site-packages/liftoff/extract_features.py", line 17, in extract_features_to_lift
    feature_db = create_feature_db_connections(args)
  File "/home/usr/anaconda3/envs/liftoff/lib/python3.6/site-packages/liftoff/extract_features.py", line 32, in create_feature_db_connections
    feature_db = build_database(args.db, args.g, disable_transcripts, disable_genes)
  File "/home/usr/anaconda3/envs/liftoff/lib/python3.6/site-packages/liftoff/extract_features.py", line 47, in build_database
    return feature_db
UnboundLocalError: local variable 'feature_db' referenced before assignment 

My command is 
liftoff /home/usr/genome/seq/PTRv2_chrxm.fa /home/usr/genome/seq/GRCh38p12_chrxm.fa -g gencode.GRCh38p13_chrxm_wo_header.gff3 -o lifted_off.panTro6_chrxm.gff3 -u NOT_liftedoff.panTro6_chrxm.gff3  -p 32 2>&1 1> chimp_liftoff.log

My liftoff version is v1.6.1

Thank you in advance,

**Comment by agshumate** [comment]
Hi,
I apologize for my slow response. As mentioned in other issues, the error handling here needs to be improved. but generally this is indicative of a problem with the gff3 file (or permissions which you seem to have ruled out). I'd be happy to take a look at your file, but if you cant share it, i suggest running AGAT (https://github.com/NBISweden/AGAT) to sanitize your file and clean up anything that might be causing the database build to fail.

**Comment by maximilianpress** [comment]
@Kohei-Fujiwara89 One thing that helped me figure it out was to run liftoff's gffutils command interactively to see if a more informative error shows up:

```
> import gffutils
> featuredb = gffutils.create_db("test.gff", "fake_db", merge_strategy="create_unique", force=True, disable_infer_transcripts=True, disable_infer_genes=True, verbose=True)
```

**Comment by Kohei-Fujiwara89** [comment]
@agshumate 
Thank you for Reply! The file is here: [http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_38/gencode.v38.annotation.gtf.gz](url)
or this gff3[http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_38/gencode.v38.annotation.gff3.gz](url).
I just downloaded them, followed by the above command.
Also, I ran AGAT like this: 
`agat_convert_sp_gxf2gxf.pl -g /home/usr/genome/gtf/gencode.v38.annotation.gtf -o /home/usr/genome/gtf/sanitized_gencode.v38.annotation.gff3`
and used the gotten file to liftoff, again failed to the same error.

Perhaps it may be common in using gffutils?

**Comment by Kohei-Fujiwara89** [comment]
@maximilianpress 
Thank you for your suggestion. 
I did them `gffutils.create_db("sanitized_gencode.v38.annotation.gff3", "sanitized_gencode.v38.annotation_db", merge_strategy='create_unique', force=True, disable_infer_transcripts=True, disable_infer_genes=True, verbose=True)`
the error messages are this:
`Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/usr/anaconda3/envs/gffutils/lib/python3.6/site-packages/gffutils/create.py", line 1292, in create_db
    c.create()
  File "/home/usr/anaconda3/envs/gffutils/lib/python3.6/site-packages/gffutils/create.py", line 506, in create
    self._init_tables()
  File "/home/usr/anaconda3/envs/gffutils/lib/python3.6/site-packages/gffutils/create.py", line 438, in _init_tables
    self.set_pragmas(self.pragmas)
  File "/home/usr/anaconda3/envs/gffutils/lib/python3.6/site-packages/gffutils/create.py", line 427, in set_pragmas
    ['PRAGMA %s=%s' % i for i in self.pragmas.items()]
sqlite3.OperationalError: database is locked
`
![image](https://user-images.githubusercontent.com/79632070/132162707-658b164f-c5f9-478b-9729-d0c0c649112f.png)
Is it because the permission of new file be -rw-r--r-- ?

And the above error seems common when using gffutils: `https://www.google.com/search?q=gffutils+create+database+is+%22locked%22+sqlite3`

**Comment by maximilianpress** [comment]
@Kohei-Fujiwara89 The permissions may indeed be an issue. Recall that you are trying to write to that file, so changing it so users can write might fix it. Alternately, you could just move your GFF to a different directory with write permissions and it might work. 

liftoff is simply taking your GFF3 file path, removing `.gff3`, and writing the db there. So if you move it to a directory with write permissions the db will show up in the same directory.

If it turns out to not be permissions (or not only permissions), there is a helpful SO page [here.](https://stackoverflow.com/questions/2740806/python-sqlite-database-is-locked)

**Comment by Kohei-Fujiwara89** [comment]
@maximilianpress 
Thank you for useful advice. Following your advice, I changed my mind to run liftoff locally, instead of running in my lab's data server. Finally it ran successfully, and it took just some hours!
I thought transferring annotation was such a heavy task that I cannot run it locally without gorgeous performance.
Also, I found that the errors are due to the settings of our lab server, which prohibits modifying one file simultaneously by others.
Thank you so much!

@agshumate 
Thank you so much for taking time.
It's wonderful tool, solving poor-annotation-problems of non-model species!
